torchtitan/experiments/deepseek_v3/generate.py

```diff
@@ -181,14 +181,51 @@ def decode(tokenizer, x):
     return colored_output
 
 
-@torch.inference_mode()
+def time_generation(func):
+    """Decorator to time generation functions and display timing and token per second results."""
+
+    @torch.inference_mode()
+    def wrapper(*args, use_timer=True, **kwargs):
+        rank = dist.get_rank()
+
+        # Setup timer if requested
+        if use_timer:
+            torch.cuda.synchronize()
+            start_event = torch.cuda.Event(enable_timing=True)
+            end_event = torch.cuda.Event(enable_timing=True)
+            start_event.record()
+
+        # Call the original function
+        result = func(*args, **kwargs)
+
+        # Record end time and calculate elapsed time
+        if use_timer:
+            end_event.record()
+            torch.cuda.synchronize()
+            elapsed_time = start_event.elapsed_time(end_event)
+            n_tokens = kwargs.get("n_tokens", 200)
```

Comment:
nit: maybe make the `generate` function return number of tokens actually generated?

Reply:
great point!  I've updated the generate function to track total tokens generated and pass that back so now we have proper tokens_per_second.  
Earlier, I was using the n_tokens setting and right now generation was just running to the full n_tokens, regardless of having already generated an eos_token. 
I've updated it to do an on device (no cpu sync) check for eos_token and now it both displays the total tokens generated for the prompt and has the correct tokens_per_second even in earlier finish. 
Anyway, now it's displaying total tokens generated and ensures that tokens_per_second is always accurate.